### PR TITLE
Implement individual CSS transform properties: translate, rotate, scale

### PR DESCRIPTION
**Pull request by birtles**
These are the servo-side changes for [Gecko bug 1207734](https://bugzilla.mozilla.org/show_bug.cgi?id=1207734).

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix [Gecko bug 1207734](https://bugzilla.mozilla.org/show_bug.cgi?id=1207734)
- [X] There are tests for these changes on the Gecko side including web-platform-tests

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19859)

### CODE REVIEW

**Comment by birtles** [comment]
These patches were written by @CJKu. I am just landing them on his behalf. They have been reviewed by @emilio and myself.

I'll wait for the first part of the Gecko-side patches to be landed and merged to autoland before landing this (and I'll update the Gecko bindings at that time). I'm opening this for now so others can have a look and in case the automated tests pick up anything.

**Comment by birtles** [comment]
Failure was:

> error: failed to execute compile
> caused by: error reading compile response from server
> caused by: Failed to read response header
> caused by: failed to fill whole buffer
> error: Could not compile `script`.

**Comment by birtles** [comment]
@bors-servo retry

* infra

**Comment by birtles** [comment]
As for the Taskcluster failure, the [log](https://public-artifacts.taskcluster.net/EP2crE3RRIadSc_5scuT5A/0/public/logs/live_backing.log) just has:
> It looks like rustup is not installed. See instructions at https://github.com/servo/servo/#setting-up-your-environment

?

**Comment by emilio** [comment]
On 01/25/2018 07:45 AM, Brian Birtles wrote:
> As for the Taskcluster failure, the log
> <https://public-artifacts.taskcluster.net/EP2crE3RRIadSc_5scuT5A/0/public/logs/live_backing.log>
> just has:
> 
>     It looks like rustup is not installed. See instructions at
>     https://github.com/servo/servo/#setting-up-your-environment
> 
> ?

Yeah, taskcluster doesn't block landing yet, just ignore that.

See:
https://groups.google.com/d/msg/mozilla.dev.servo/lttaOHEMzes/SYTcTk88AAAJ

**Comment by jdm** [comment]
```
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "e.style['translate'] = \"100deg\" should not set the property value", "test": "/css/css-transforms/parsing/translate-parsing-invalid.html", "line": 127937, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "e.style['translate'] = \"100px 200px 300%\" should not set the property value", "test": "/css/css-transforms/parsing/translate-parsing-invalid.html", "line": 127938, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "e.style['translate'] = \"100px 200px calc(30px + 30%)\" should not set the property value", "test": "/css/css-transforms/parsing/translate-parsing-invalid.html", "line": 127939, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "e.style['scale'] = \"100px\" should not set the property value", "test": "/css/css-transforms/parsing/scale-parsing-invalid.html", "line": 135858, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "e.style['scale'] = \"100 200 300 400\" should not set the property value", "test": "/css/css-transforms/parsing/scale-parsing-invalid.html", "line": 135859, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "\"100deg\" and \"180deg\" are valid rotate values", "test": "/css/css-transforms/animation/rotate-interpolation.html", "line": 139961, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "\"45deg\" and \"-1 1 0 60deg\" are valid rotate values", "test": "/css/css-transforms/animation/rotate-interpolation.html", "line": 139968, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "\"none\" and \"7 -8 9 400grad\" are valid rotate values", "test": "/css/css-transforms/animation/rotate-interpolation.html", "line": 139975, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "\"none\" and \"none\" are valid rotate values", "test": "/css/css-transforms/animation/rotate-interpolation.html", "line": 139982, "action": "test_result", "expected": "FAIL"}
{"status": "FAIL", "group": "default", "message": "assert_equals: expected \"calc(10px + 10%) calc(20px + 20%) calc(30px + 30em)\" but got \"calc(10% + 10px) calc(20% + 20px) calc(30em + 30px)\"", "stack": "test_valid_value/<@http://web-platform.test:8000/css/css-transforms/parsing/resources/parsing-testcommon.js:19:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1494:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\ntest_valid_value@http://web-platform.test:8000/css/css-transforms/parsing/resources/parsing-testcommon.js:15:5\n@http://web-platform.test:8000/css/css-transforms/parsing/translate-parsing-valid.html:26:1\n", "subtest": "Serialization should round-trip after setting e.style['translate'] = \"calc(10px + 10%) calc(20px + 20%) calc(30px + 30em)\"", "test": "/css/css-transforms/parsing/translate-parsing-valid.html", "line": 140039, "action": "test_result", "expected": "PASS"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Serialization should round-trip after setting e.style['translate'] = \"0\"", "test": "/css/css-transforms/parsing/translate-parsing-valid.html", "line": 140041, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Serialization should round-trip after setting e.style['translate'] = \"1px 2px 0\"", "test": "/css/css-transforms/parsing/translate-parsing-valid.html", "line": 140043, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "e.style['rotate'] = \"100px\" should not set the property value", "test": "/css/css-transforms/parsing/rotate-parsing-invalid.html", "line": 143951, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "e.style['rotate'] = \"100 400deg\" should not set the property value", "test": "/css/css-transforms/parsing/rotate-parsing-invalid.html", "line": 143952, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "e.style['rotate'] = \"100 200 400deg\" should not set the property value", "test": "/css/css-transforms/parsing/rotate-parsing-invalid.html", "line": 143953, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "e.style['rotate'] = \"100 200 300 500 400deg\" should not set the property value", "test": "/css/css-transforms/parsing/rotate-parsing-invalid.html", "line": 143954, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "\"2 30 400\" and \"10 110 1200\" are valid scale values", "test": "/css/css-transforms/animation/scale-interpolation.html", "line": 149210, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "\"26 17 9\" and \"2\" are valid scale values", "test": "/css/css-transforms/animation/scale-interpolation.html", "line": 149217, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "\"none\" and \"4 3 2\" are valid scale values", "test": "/css/css-transforms/animation/scale-interpolation.html", "line": 149224, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "\"none\" and \"none\" are valid scale values", "test": "/css/css-transforms/animation/scale-interpolation.html", "line": 149231, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "\"220px 240px 260px\" and \"300px 400px 500px\" are valid translate values", "test": "/css/css-transforms/animation/translate-interpolation.html", "line": 149256, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "\"480px 400px 320px\" and \"240% 160%\" are valid translate values", "test": "/css/css-transforms/animation/translate-interpolation.html", "line": 149263, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "\"none\" and \"8px 80% 800px\" are valid translate values", "test": "/css/css-transforms/animation/translate-interpolation.html", "line": 149270, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "\"none\" and \"none\" are valid translate values", "test": "/css/css-transforms/animation/translate-interpolation.html", "line": 149277, "action": "test_result", "expected": "FAIL"}
```

File: components/style/properties/helpers/animated_properties.mako.rs
Comment (nox) [inline components/style/properties/helpers/animated_properties.mako.rs]: Why is this call not like the others? Can't the type of `self.quaternion` implement `Animate` and use `MatrixDecomposed3D::slerp` there? That would make this implementation for `MatrixDecomposed3D` derivable.

File: components/style/properties/helpers/animated_properties.mako.rs
Comment (nox) [inline components/style/properties/helpers/animated_properties.mako.rs]: Indentation is broken. Spacing is broken.

File: components/style/properties/gecko.mako.rs
Comment (nox) [inline components/style/properties/gecko.mako.rs]: Nit: `if let` makes for more readable code here.

File: components/style/properties/helpers/animated_properties.mako.rs
Comment (nox) [inline components/style/properties/helpers/animated_properties.mako.rs]: Make this an implementation of `Animatable` for `Quaternion` instead of keeping it on `MatrixDecomposed3D` .

File: components/style/properties/helpers/animated_properties.mako.rs
Comment (nox) [inline components/style/properties/helpers/animated_properties.mako.rs]: I don't understand the relation between that method name and the spec link just above.

File: components/style/properties/helpers/animated_properties.mako.rs
Comment (nox) [inline components/style/properties/helpers/animated_properties.mako.rs]: Nit: it's more readable to instead dereference the matched value `*translate` and not prefix every pattern with `&`.

**Review by nox — CHANGES_REQUESTED** [review]

**Comment by birtles** [comment]
Thanks for the review!
Please bear in mind these are not my patches -- I'm just landing them so I can't answer all the questions. I'll make requested changes next week though.

**Comment by nox** [comment]
@birtles ACK, didn't realise that! I think we can do follow-up issues if you don't have the bandwidth to make the changes.

**Comment by birtles** [comment]
Thanks! I'll try to get to it. I think there were some test failures anyway that I need to look into anyway.

**Comment by birtles** [comment]
> assert_equals: expected \"calc(10px + 10%) calc(20px + 20%) calc(30px + 30em)\" but got \"calc(10% + 10px) calc(20% + 20px) calc(30em + 30px)\""

Looks like we serialize the percent first instead of the number but the [spec says numbers go first](https://drafts.csswg.org/css-values/#math-function-serialize-a-summation).

**Comment by birtles** [comment]
> Looks like we serialize the percent first instead of the number but the spec says numbers go first.

Oh, no wait, we have a length here so it should go after the percent. So it looks like the test is wrong.

**Comment by birtles** [comment]
I've rebased the PR and addressed half of the review feedback (not the quarternion part yet). I've also yet to fix the failing wpt.

**Comment by birtles** [comment]
Added https://github.com/w3c/web-platform-tests/pull/9292 to fix web platform test.

**Comment by birtles** [comment]
All the review feedback should now be addressed.

**Comment by birtles** [comment]
@jdm I fixed the failing wpt upstream. Do you know if anyone is planning on doing a wpt sync soon?

**Comment by emilio** [comment]
@birtles you can update it manually here, it should be fine

**Comment by jdm** [comment]
https://github.com/servo/servo/pull/19913

**Comment by birtles** [comment]
> #19913

Thank you!

**Comment by birtles** [comment]
I have no idea why the commits appear out of order here :/

File: components/style/values/specified/transform.rs
Comment (emilio) [inline components/style/values/specified/transform.rs]: nit: I pointed out on the original review but apparently only some of them were fixed. There's no need to do `.ok()` here, you can just do:

```rust
if let Ok(sy) = input.try(|i| Number::parse(context, i)) {
```

And ditto below.

File: components/style/properties/helpers/animated_properties.mako.rs
Comment (emilio) [inline components/style/properties/helpers/animated_properties.mako.rs]: nit: There's a missing space after `>` here.

File: components/style/properties/helpers/animated_properties.mako.rs
Comment (emilio) [inline components/style/properties/helpers/animated_properties.mako.rs]: Ditto here.

File: components/style/properties/helpers/animated_properties.mako.rs
Comment (emilio) [inline components/style/properties/helpers/animated_properties.mako.rs]: I wonder if this should be done during `ToComputedValue` instead of preserving the `Option`s and variants at computed-value time... I guess it's not a big deal either way.

**Review by emilio — COMMENTED** [review]
GH used to put the commits in author order? huh.

File: components/style/properties/helpers/animated_properties.mako.rs
Comment (birtles) [reply components/style/properties/helpers/animated_properties.mako.rs]: Yeah, I don't have time to do any more work on these patches. I think we can do this in a follow up.

**Comment by birtles** [comment]
Hmm, now `./mach build-geckolib` gives me:

> error[E0609]: no field `mBoolFlags` on type `gecko_bindings::structs::root::nsWrapperCache`
>    --> components/style/gecko/wrapper.rs:194:32
>     |
> 194 |         (self.0)._base._base_1.mBoolFlags
>     |                                ^^^^^^^^^^ unknown field
>     |
>     = note: available fields are: `vtable_`, `mWrapper`, `mFlags`

I guess there is some mismatch in the setting of `BOOL_FLAGS_ON_WRAPPER_CACHE` between the `structs.rs` I updated in this PR (which I just took off emilio's build on autoland) and what we previously had.

**Comment by birtles** [comment]
Oh, I probably mistakenly took the 32-bit bindings.

**Comment by emilio** [comment]
r=me, when you want.

@bors-servo delegate+

**Comment by birtles** [comment]
Thanks. I'll wait for #19913 to land and then rebase.

**Comment by birtles** [comment]
Oh, I didn't realize the PASSes that @jdm originally quoted were actually failures!
I guess there is an expectations file somewhere I need to update.

**Pull request merged**
